from snakemake.utils import min_version

min_version("7.30.1")

SAMPLES = []
SAMPLES_BY_CELL = []


rule all:
    input:
        expand(
            f"results/{samples}_quant_bambu/counts_transcript.txt", samples=SAMPLES_BY_CELL
        )

rule discovery_bambu:
    input:
        reads=expand("results/{samples}.aligned.sorted.bam", samples=SAMPLES),
        indices=expand("results/{samples}.aligned.sorted.bam.bai", samples=SAMPLES),
        annotation="data-raw/human_GRCh38_no_alt_plus_SIRV4.gtf",
        reference="data-raw/human_GRCh38_no_alt_plus_SIRV4.fasta",
    params:
        discovery=True,
        output_dir="results/discovery_bambu"
    output:
        extended_annotations="results/discovery_bambu/extended_annotations.gtf"
    singularity:
        "docker://condaforge/mambaforge:23.1.0-3"
    conda:
        "envs/bambu.yaml"
    script:
        "scripts/run_bambu.R"

rule quant_bambu:
    input:
        reads="results/{samples}.bycell.aligned.sorted.bam",
        indices="results/{samples}.bycell.aligned.sorted.bam.bai",
        annotation="results/discovery_bambu/extended_annotations.gtf"
        reference="data-raw/human_GRCh38_no_alt_plus_SIRV4.fasta",
    params:
        discovery=False,
        output_dir=lambda wc: f"results/{wc.samples}_quant_bambu"
    output:
        quants="results/{samples}_quant_bambu/counts_transcript.txt"
    singularity:
        "docker://condaforge/mambaforge:23.1.0-3"
    conda:
        "envs/bambu.yaml"
    script:
        "scripts/run_bambu.R"
